Indexing/LiteralIndex.hpp

```diff
@@ -25,121 +25,123 @@
 
 namespace Indexing {
 
+template<class Data>
 class LiteralIndex
 : public Index
 {
 public:
-  virtual ~LiteralIndex();
+  VirtualIterator<LiteralClause> getAll()
+  { return _is->getAll(); }
 
-  SLQueryResultIterator getAll();
-
-  SLQueryResultIterator getUnifications(Literal* lit, bool complementary, bool retrieveSubstitutions = true)
+  VirtualIterator<QueryRes<ResultSubstitutionSP, LiteralClause>> getUnifications(Literal* lit, bool complementary, bool retrieveSubstitutions = true)
   { return _is->getUnifications(lit, complementary, retrieveSubstitutions); }
 
-  VirtualIterator<LQueryRes<AbstractingUnifier*>> getUwa(Literal* lit, bool complementary, Options::UnificationWithAbstraction uwa, bool fixedPointIteration)
+  VirtualIterator<QueryRes<AbstractingUnifier*, Data>> getUwa(Literal* lit, bool complementary, Options::UnificationWithAbstraction uwa, bool fixedPointIteration)
   { return _is->getUwa(lit, complementary, uwa, fixedPointIteration); }
 
-  SLQueryResultIterator getGeneralizations(Literal* lit,
-	  bool complementary, bool retrieveSubstitutions = true);
+  VirtualIterator<QueryRes<ResultSubstitutionSP, LiteralClause>> getGeneralizations(Literal* lit, bool complementary, bool retrieveSubstitutions = true)
+  { return _is->getGeneralizations(lit, complementary, retrieveSubstitutions); }
 
-  SLQueryResultIterator getInstances(Literal* lit,
-	  bool complementary, bool retrieveSubstitutions = true);
+  VirtualIterator<QueryRes<ResultSubstitutionSP, LiteralClause>> getInstances(Literal* lit, bool complementary, bool retrieveSubstitutions = true)
+  { return _is->getInstances(lit, complementary, retrieveSubstitutions); }
 
-  size_t getUnificationCount(Literal* lit, bool complementary);
+  size_t getUnificationCount(Literal* lit, bool complementary)
+  { return _is->getUnificationCount(lit, complementary); }
 
 
   friend std::ostream& operator<<(std::ostream& out,                 LiteralIndex const& self) { return out << *self._is; }
   friend std::ostream& operator<<(std::ostream& out, OutputMultiline<LiteralIndex>const& self) { return out << multiline(*self.self._is, self.indent); }
 
 protected:
-  LiteralIndex(LiteralIndexingStructure* is) : _is(is) {}
+  LiteralIndex(LiteralIndexingStructure<Data>* is) : _is(is) {}
 
-  void handleLiteral(Literal* lit, Clause* cl, bool add);
+  void handle(Data data, bool add)
+  { _is->handle(std::move(data), add); }
```

Comment:
Could one make it explicit, with `via handle(Data&& data, bool add)` that data is going to be moved? Or was the intent to copy data to `LiteralIndex::handle` but move them to `_is->handle`?)

Reply:
Yes exactly this is the reason why we're using the tie here. We can't do this in `LiteralClause` because we can't return a reference to the result of `clause->number()` and `literal->getId()` (as they are only live during the call of `asTuple`), thus we need to use `make_tuple` in this case. In any case we want as tuple just to return some tuple of which evey element can be ordered using `std::less`, hashed, ..., and what exact the type tuple elements are doesn't matter.

Reply:
I was just guessing that since you currently only use `handle` with temporaries, we might get a tiny more efficient code with `&&`. Or is this a place where the compiler can save the copy for us, @MichaelRawson ?

Reply:
I think in terms of efficiency it should be just the same as long as we didn't implement the constructor `Data(Data&&)` in some unexpected way. For me the design-question would just be whether we want to force the caller to explicitly call a copy constructor or not.

Reply:
Temporaries are probably OK, I think this is more a matter of taste as Joe suggests. I expect most are now guaranteed to be elided since C++17, see the example at the bottom of https://en.cppreference.com/w/cpp/language/copy_elision

As a rule I'd probably take everything by value if you need to own the object (unless you semantically wanted an rvalue for some reason), allowing either a copy or move. This may be shaped by Rust habits more than C++ considerations.